overlord/devicestate/handlers_install.go

```diff
@@ -691,6 +692,77 @@ func readPreseedAssertion(st *state.State, model *asserts.Model, ubuntuSeedDir,
 
 var seedOpen = seed.Open
 
+// TODO: consider reusing this kind of handler for UC20 seeding
+type preseedSnapHandler struct {
+	writableDir string
+}
+
+func (p *preseedSnapHandler) HandleUnassertedSnap(name, path string, _ timings.Measurer) (string, error) {
+	pinfo := snap.MinimalPlaceInfo(name, snap.Revision{N: -1})
+	targetPath := filepath.Join(p.writableDir, pinfo.MountFile())
+	mountDir := filepath.Join(p.writableDir, pinfo.MountDir())
+
+	sq := squashfs.New(path)
+	opts := &snap.InstallOptions{MustNotCrossDevices: true}
+	if _, err := sq.Install(targetPath, mountDir, opts); err != nil {
+		return "", fmt.Errorf("cannot install snap %q: %v", name, err)
+	}
+
+	return targetPath, nil
+}
+
+func (p *preseedSnapHandler) HandleAndDigestAssertedSnap(name, path string, essType snap.Type, snapRev *asserts.SnapRevision, _ func(string, uint64) (snap.Revision, error), _ timings.Measurer) (string, string, uint64, error) {
+	pinfo := snap.MinimalPlaceInfo(name, snap.Revision{N: snapRev.SnapRevision()})
+	targetPath := filepath.Join(p.writableDir, pinfo.MountFile())
+	mountDir := filepath.Join(p.writableDir, pinfo.MountDir())
+
+	logger.Debugf("copying: %q to %q; mount dir=%q", path, targetPath, mountDir)
+
+	srcFile, err := os.Open(path)
+	if err != nil {
+		return "", "", 0, err
+	}
+	defer srcFile.Close()
+
+	destFile, err := osutil.NewAtomicFile(targetPath, 0644, 0, osutil.NoChown, osutil.NoChown)
+	if err != nil {
+		return "", "", 0, fmt.Errorf("cannot create atomic file: %v", err)
+	}
+	defer destFile.Cancel()
+
+	finfo, err := srcFile.Stat()
+	if err != nil {
+		return "", "", 0, err
+	}
+
+	destFile.SetModTime(finfo.ModTime())
+
+	h := crypto.SHA3_384.New()
+	w := io.MultiWriter(h, destFile)
+
+	size, err := io.CopyBuffer(w, srcFile, make([]byte, 2*1024*1024))
```

Comment:
(nitpick) Reading the io.CopyBuffer documentation it seems to be fine to just pass `nil` here - unless there is a specific reason to create the buffer here (fwiw, go uses a different though).

Reply:
Yeah, sorry, I "borrowed" this from `osutil/digest.go` (hashDigestBufSize), this should probably be const as well (unfortunately the original code doesn't provide any insight about the reason behind this value).